pkg/pod/admission/nodeenv/admission.go

```diff
@@ -0,0 +1,97 @@
+package admission
+
+import (
+	"fmt"
+	"io"
+
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/admission"
+	kapi "github.com/GoogleCloudPlatform/kubernetes/pkg/api"
+	apierrors "github.com/GoogleCloudPlatform/kubernetes/pkg/api/errors"
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/api/meta"
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/client"
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/labels"
+
+	nodecache "github.com/openshift/origin/pkg/node/cache"
+	projectcache "github.com/openshift/origin/pkg/project/cache"
+)
+
+func init() {
+	admission.RegisterPlugin("OriginPodNodeEnvironment", func(client client.Interface, config io.Reader) (admission.Interface, error) {
+		return NewPodNodeEnvironment()
+	})
+}
+
+// podNodeEnvironment is an implementation of admission.Interface.
+type podNodeEnvironment struct {
+}
+
+// Admit enforces that pod and its project node label selectors matches at least a node in the cluster.
+func (p *podNodeEnvironment) Admit(a admission.Attributes) (err error) {
+	// ignore deletes
+	if a.GetOperation() == "DELETE" {
+		return nil
+	}
+
+	resource := a.GetResource()
+	if resource != "pods" {
+		return nil
+	}
+
+	obj := a.GetObject()
+	name := "Unknown"
+	if obj != nil {
+		name, _ = meta.NewAccessor().Name(obj)
+	}
+	pod := obj.(*kapi.Pod)
+
+	var podNodeSelector labels.Selector
+	if len(pod.Spec.NodeSelector) == 0 {
+		podNodeSelector = labels.Everything()
+	} else {
+		podNodeSelector = labels.SelectorFromSet(pod.Spec.NodeSelector)
+	}
+
+	projects, err := projectcache.GetProjectCache()
+	if err != nil {
+		return err
+	}
+	namespace, err := projects.GetNamespaceObject(a.GetNamespace())
+	if err != nil {
+		return apierrors.NewForbidden(resource, name, err)
+	}
+	projectNodeSelector, err := projects.GetNodeSelectorObject(namespace)
+	if err != nil {
+		return apierrors.NewForbidden(resource, name, err)
+	}
+
+	nc, err := nodecache.GetNodeCache()
+	if err != nil {
+		return err
+	}
+	nodes, err := nc.NodeLister.List()
+	if err != nil {
+		return apierrors.NewForbidden(resource, name, fmt.Errorf("Unable to list nodes in the cluster, error: %v", err))
+	}
+
+	projectMatched := false
+	for _, node := range nodes.Items {
+		nodeLabelSet := labels.Set(node.Labels)
+		podSelectorMatches := podNodeSelector.Matches(nodeLabelSet)
+		projectSelectorMatches := projectNodeSelector.Matches(nodeLabelSet)
+		if podSelectorMatches && projectSelectorMatches {
+			return nil
+		}
+		if projectSelectorMatches {
+			projectMatched = true
+		}
+	}
+	if !projectMatched {
+		return apierrors.NewForbidden(resource, name, fmt.Errorf("Pod's project node selector didn't match any nodes in the cluster."))
+	} else {
+		return apierrors.NewForbidden(resource, name, fmt.Errorf("Pod node selector didn't match any nodes in the cluster."))
```

Comment:
I don't get why you would reject the pod creation request.

I think we should accept the pod creation request, but the scheduler should error when it attempts to find a placement for the pod.  The error should be made known to the user via an event on the pod like all other pod scheduling events.

This seems to introduce a lot of overhead on every pod creation operation, when we should be looking to do this level of validation at pod placement.

Can you explain why we feel the need to reject the pod outright rather than just not schedule?